### PR TITLE
Implement mod(z::Dual, n::Number)

### PR DESCRIPTION
**Pull request by tlycken**
I just made this work on the real part and leave the dual part as-is, because that's what I need in my application. I guess that should be OK at least in some aspect, since anywhere `mod` is applicable, periodicity is implied which should make the derivative the same, but if you have better ideas, let me know :)

### CODE REVIEW

**Comment by tlycken** [comment]
Oh, I should add a couple of tests. Brb...

**Comment by mlubin** [comment]
Could you give an example where ``mod`` of a dual number makes sense?

**Comment by tlycken** [comment]
Sure: [right here](https://github.com/tlycken/interpolations.jl/pull/50/files#diff-bb1f05269422240d7948d66ed4ed80b0R7).

It might not be so straightforward just from the code, but the application is interpolation of data, in this case with periodic extrapolation (we use `mod` for reflecting extrapolation too). We've taken care to allow for indexing with anything that `isa(x, Number)`, and dual numbers as indices have a very real use case for evaluating both the interpolant and its approximate derivative (i.e. if the interpolant is a quadratic spline, the derivative is a piece-wise linear function) simultaneously.

Right before we actually forward the indices to the wrapped arrays, we use convert `real(x)` to an integer to avoid `InexactError`s, but there can be quite a few things happening before we reach that point. `DualNumbers` have been sort-of our guinea pig for trying out that this all works (mainly because we can then also use it to test our implementation of the `gradient` function...) so there might be lots of other `Number` types that we think we support but don't - currently, the spec for what works is this:

* Inherits `Number`
* Implements `+`, `-`, `*` and `/`
* Promotes nicely with at least real numbers in Base
* Implements `mod` (which is needed specifically for periodic and reflecting extrapolation)

**Comment by tlycken** [comment]
@mlubin Any further comments on this?

File: test/dual_n.jl [hunk 1]
```diff
@@ -86,3 +86,6 @@ sq = squareroot(Dual4(10000.0,1.0,2.0,3.0,4.0))
 @test_approx_eq epsilon2(NaNMath.pow(exp(1),Dual4(1.0,1.0,2.0,3.0,4.0))) 2exp(1)
 @test_approx_eq epsilon3(NaNMath.pow(exp(1),Dual4(1.0,1.0,2.0,3.0,4.0))) 3exp(1)
 @test_approx_eq epsilon4(NaNMath.pow(exp(1),Dual4(1.0,1.0,2.0,3.0,4.0))) 4exp(1)
+
+@test real(mod(dual(15.23, 1), 10)) == 5.23
```
Comment (mlubin) [inline test/dual_n.jl]: The tests are not quite well organized, but this should go into the ``automatic_differentiation_test.jl`` file.

Comment (tlycken) [inline test/dual_n.jl]: Done! Do you want me to re-squash the commits?

**Comment by mlubin** [comment]
Squashing would be appreciated

**Comment by tlycken** [comment]
Done.

**Comment by tlycken** [comment]
Thanks a lot for merging! It would be extra awesome if you were able to tag a new version as well, to allow the tests in tlycken/Interpolations.jl#50 to pass without me having to modify my build script to check out the `master` branch of DualNumbers.jl :)

**Comment by mlubin** [comment]
Done. CC @jrevels for port to ForwardDiff

**Pull request merged**